src/swarm/neo/request/RequestEventDispatcher.d

```diff
@@ -0,0 +1,777 @@
+                return true;
+
+        return false;
+    }
+
+    /***************************************************************************
+
+        Pops the first fiber awaiting a send from the list of waiting fibers.
+        Should only be called after waitingWriters() returns true.
+
+        Returns:
+            the first fiber awaiting a send from the list of waiting fibers
+
+    ***************************************************************************/
+
+    private WaitingFiber popWaitingWriter ( )
+    in
+    {
+        assert(this.waiting_fibers.length > 0);
+    }
+    out ( const_waiting_fiber )
+    {
+        auto waiting_fiber = cast(WaitingFiber)const_waiting_fiber;
+        assert(waiting_fiber.event.active == waiting_fiber.event.active.send);
+    }
+    body
+    {
+        foreach ( i, waiting_fiber; this.waiting_fibers )
+            if ( waiting_fiber.event.active == waiting_fiber.event.active.send )
+            {
+                WaitingFiber ret = waiting_fiber;
+                Array.removeShift(this.waiting_fibers, i);
+                return ret;
+            }
+
+        assert(false, "popWaitingWriter() should not be called if waitingWriters() is false");
+    }
+
+    /***************************************************************************
+
+        Dispatches any signals which occurred while the request-on-conn fiber
+        was running. These are queued up in this.queue_signals.
+
+        Params:
+            conn = request-on-conn event dispatcher (require by dispatchSignal)
+
+    ***************************************************************************/
+
+    private void dispatchQueuedSignals ( RequestOnConnBase.EventDispatcher conn )
+    out
+    {
+        assert(this.queued_signals.length == 0);
+    }
+    body
+    {
+        while ( this.queued_signals.length > 0 )
+        {
+            auto code = this.queued_signals[0];
+            Array.removeShift(this.queued_signals, 0);
+            this.dispatchSignal(conn, code);
+        }
+    }
+
+    /***************************************************************************
+
+        Helper function for eventLoop(). Dispatches the specified signal to the
+        appropriate waiting fiber.
+
+        Params:
+            conn = request-on-conn event dispatcher (used to throw a protocol
+                error, if the signal is not handled)
+            signal = signal code
+
+        Throws:
+            if the signal is not handled by any waiting fiber
+
+    ***************************************************************************/
+
+    private void dispatchSignal ( RequestOnConnBase.EventDispatcher conn,
+        int signal )
+    {
+        if ( signal < 0 ) // Standard send/receive codes
+            return;
+
+        assert(signal <= 255);
+        ubyte signal_ubyte = cast(ubyte)signal;
+
+        foreach ( waiting_fiber; this.waiting_fibers )
```

Comment:
One note here on the performance side (not sure if it's worth paying attention now, but it could be useful if it turns out that these loops are very tight when handling a lot of requests): Instead of a single list `this.waiting_fibers` we could have `this.fibers_waiting_for_message` and `this.fibers_waiting_for_signal` and then even these could actually be arrays of arrays, where the first dimension is the message/signal code, so you can access the first waiting fiber in constant time.